File: services/backend/pkg/models/status_pages.go
package models

import (
	"time"

	"github.com/google/uuid"
	"github.com/uptrace/bun"
)

const (
	StatusPageVisibilityPrivate       = "private"
	StatusPageVisibilityPublic        = "public"
	StatusPageVisibilityAuthenticated = "authenticated"
)

type StatusPage struct {
	bun.BaseModel `bun:"table:status_pages"`

	ID                   uuid.UUID                       `bun:",pk,type:uuid,default:gen_random_uuid()" json:"id"`
	Name                 string                          `bun:"name,type:text,notnull" json:"name"`
	Slug                 string                          `bun:"slug,type:text,notnull,unique" json:"slug"`
	Description          string                          `bun:"description,type:text,default:''" json:"description"`
	Visibility           string                          `bun:"visibility,type:varchar(20),notnull,default:'private'" json:"visibility"`
	IncludeAllTags       bool                            `bun:"include_all_tags,type:boolean,notnull,default:false" json:"include_all_tags"`
	ImagePatterns        StringList                      `bun:"image_patterns,type:jsonb,notnull,default:'[]'" json:"image_patterns"`
	StaleAfterHours      int                             `bun:"stale_after_hours,type:int,notnull,default:72" json:"stale_after_hours"`
	OwnerType            string                          `bun:"owner_type,type:text,notnull,default:'user'" json:"owner_type"`
	OwnerUserID          *uuid.UUID                      `bun:"owner_user_id,type:uuid" json:"owner_user_id,omitempty"`
	OwnerOrgID           *uuid.UUID                      `bun:"owner_org_id,type:uuid" json:"owner_org_id,omitempty"`
	CreatedAt            time.Time                       `bun:"created_at,type:timestamptz,default:now()" json:"created_at"`
	UpdatedAt            time.Time                       `bun:"updated_at,type:timestamptz" json:"updated_at"`
	Targets              []StatusPageTarget              `bun:"rel:has-many,join:id=page_id" json:"targets,omitempty"`
	Updates              []StatusPageUpdate              `bun:"rel:has-many,join:id=page_id" json:"updates,omitempty"`
	GitRepositorySources []StatusPageGitRepositorySource `bun:"rel:has-many,join:id=page_id" json:"git_repository_sources,omitempty"`
}

// StatusPageGitRepositorySource makes a status page follow the current image
// inventory of a Git repository rather than a fixed image reference.
type StatusPageGitRepositorySource struct {
	bun.BaseModel `bun:"table:status_page_git_repository_sources"`

	ID           uuid.UUID      `bun:",pk,type:uuid,default:gen_random_uuid()" json:"id"`
	PageID       uuid.UUID      `bun:"page_id,type:uuid,notnull" json:"page_id"`
	RepositoryID uuid.UUID      `bun:"repository_id,type:uuid,notnull" json:"repository_id"`
	ImageNames   StringList     `bun:"image_names,type:jsonb,notnull,default:'[]'" json:"image_names"`
	DisplayOrder int            `bun:"display_order,type:int,notnull,default:0" json:"display_order"`
	CreatedAt    time.Time      `bun:"created_at,type:timestamptz,default:now()" json:"created_at"`
	Repository   *GitRepository `bun:"rel:belongs-to,join:repository_id=id" json:"repository,omitempty"`
}

type StatusPageTarget struct {
	bun.BaseModel `bun:"table:status_page_targets"`

	ID           uuid.UUID `bun:",pk,type:uuid,default:gen_random_uuid()" json:"id"`
	PageID       uuid.UUID `bun:"page_id,type:uuid,notnull" json:"page_id"`
	ImageName    string    `bun:"image_name,type:text,notnull" json:"image_name"`
	ImageTag     string    `bun:"image_tag,type:text,notnull" json:"image_tag"`
	DisplayOrder int       `bun:"display_order,type:int,notnull,default:0" json:"display_order"`
	CreatedAt    time.Time `bun:"created_at,type:timestamptz,default:now()" json:"created_at"`
}

type StatusPageUpdate struct {
	bun.BaseModel `bun:"table:status_page_updates"`

	ID              uuid.UUID  `bun:",pk,type:uuid,default:gen_random_uuid()" json:"id"`
	PageID          uuid.UUID  `bun:"page_id,type:uuid,notnull" json:"page_id"`
	Title           string     `bun:"title,type:text,notnull" json:"title"`
	Body            string     `bun:"body,type:text,default:''" json:"body"`
	Level           string     `bun:"level,type:varchar(20),notnull,default:'info'" json:"level"`
	ActiveFrom      *time.Time `bun:"active_from,type:timestamptz" json:"active_from,omitempty"`
	ActiveUntil     *time.Time `bun:"active_until,type:timestamptz" json:"active_until,omitempty"`
	CreatedByUserID uuid.UUID  `bun:"created_by_user_id,type:uuid,notnull" json:"created_by_user_id"`
	CreatedAt       time.Time  `bun:"created_at,type:timestamptz,default:now()" json:"created_at"`
	UpdatedAt       time.Time  `bun:"updated_at,type:timestamptz" json:"updated_at"`
}

File: services/frontend/lib/api/types/status-pages.ts
import type { OwnerType } from './common';
import type { BlockedPolicyDetails } from './scans';

export interface StatusPageTarget {
  id?: string;
  page_id?: string;
  image_name: string;
  image_tag: string;
  display_order: number;
  created_at?: string;
}

export interface StatusPageGitRepositorySource {
  id?: string;
  page_id?: string;
  repository_id: string;
  image_names?: string[];
  display_order: number;
  created_at?: string;
  repository?: { id: string; name: string };
}

export interface StatusPageGitRepositorySourceHealth {
  repository_id: string;
  repository_name: string;
  display_order: number;
  status: string;
  latest_run_id?: string;
  latest_run_status?: string;
  snapshot_run_id?: string;
  commit_sha?: string;
  completed_at?: string;
  image_count: number;
}

export interface StatusPageUpdate {
  id?: string;
  page_id?: string;
  title: string;
  body: string;
  level: 'info' | 'maintenance' | 'incident';
  active_from?: string | null;
  active_until?: string | null;
  created_by_user_id?: string;
  created_at?: string;
  updated_at?: string;
}

export interface StatusPage {
  id: string;
  name: string;
  slug: string;
  description: string;
  visibility: 'private' | 'public' | 'authenticated';
  include_all_tags: boolean;
  image_patterns?: string[];
  stale_after_hours: number;
  owner_type?: OwnerType;
  owner_user_id?: string | null;
  owner_org_id?: string | null;
  created_at: string;
  updated_at: string;
  targets?: StatusPageTarget[];
  updates?: StatusPageUpdate[];
  git_repository_sources?: StatusPageGitRepositorySource[];
}

export interface StatusPageItem {
  image_name: string;
  image_tag: string;
  latest_scan_id: string;
  scan_status: string;
  external_status?: string;
  compliance_status?: 'pass' | 'fail';
  scan_provider?: string;
  current_step?: string;
  started_at?: string;
  status: string;
  error_message?: string;
  blocked_policy_details?: BlockedPolicyDetails | null;
  critical_count: number;
  high_count: number;
  medium_count: number;
  low_count: number;
  previous_scan_id?: string;
  previous_critical_count?: number;
  previous_high_count?: number;
  previous_medium_count?: number;
  previous_low_count?: number;
  delta_critical_count?: number;
  delta_high_count?: number;
  delta_medium_count?: number;
  delta_low_count?: number;
  freshness_hours: number;
  observed_at: string;
  previous_scan_at?: string;
  display_order: number;
  source_type?: 'git_repository';
  source_repository_id?: string;
  source_repository_name?: string;
  source_run_id?: string;
  full_ref?: string;
  discovery_state?: string;
}

export interface StatusPageScanSummary {
  scan_id: string;
  image_name: string;
  image_tag: string;
  scan_status: string;
  external_status?: string;
  compliance_status?: 'pass' | 'fail';
  scan_provider?: string;
  current_step?: string;
  error_message?: string;
  blocked_policy_details?: BlockedPolicyDetails | null;
  critical_count: number;
  high_count: number;
  medium_count: number;
  low_count: number;
  started_at?: string;
  completed_at?: string;
  created_at: string;
  observed_at: string;
  is_latest: boolean;
}

export interface StatusPageResponse {
  page: StatusPage;
  items: StatusPageItem[];
  git_repository_sources?: StatusPageGitRepositorySourceHealth[];
  now: string;
}

export interface StatusPagePayload {
  name: string;
  slug?: string;
  description?: string;
  visibility: 'private' | 'public' | 'authenticated';
  org_id?: string;
  include_all_tags: boolean;
  image_patterns?: string[];
  stale_after_hours: number;
  targets: StatusPageTarget[];
  git_repository_sources?: StatusPageGitRepositorySource[];
  updates?: StatusPageUpdate[];
}

export interface StatusPageTargetOption {
  id: string;
  image_name: string;
  image_tag: string;
  label: string;
  latest_scan_id: string;
  latest_status: string;
  observed_at: string;
  critical_count: number;
  high_count: number;
}
